crates/ide/src/view_memory_layout.rs

```diff
@@ -0,0 +1,373 @@
+                    typename: child_ty.display(db).to_string(),
+                    size: 0,
+                    offset: 0,
+                    alignment: 0,
+                    parent_idx: parent_idx as i64,
+                    children_start: -1,
+                    children_len: 0,
+                });
+            }
+        }
+
+        for (i, (_, child_ty)) in fields.iter().enumerate() {
+            if let Ok(child_layout) = child_ty.layout(db) {
+                read_layout(nodes, db, &child_ty, &child_layout, children_start + i);
+            }
+        }
+    }
+
+    ty.layout(db)
+        .map(|layout| {
+            let item_name = match def {
+                // def is a datatype
+                Definition::Adt(_)
+                | Definition::TypeAlias(_)
+                | Definition::BuiltinType(_)
+                | Definition::SelfType(_) => "[ROOT]".to_owned(),
+
+                // def is an item
+                def => def
+                    .name(db)
+                    .map(|n| {
+                        n.as_str()
+                            .map(|s| s.to_owned())
+                            .unwrap_or_else(|| format!(".{}", n.as_tuple_index().unwrap()))
+                    })
+                    .unwrap_or("[ROOT]".to_owned()),
+            };
+
+            let typename = ty.display(db).to_string();
+
+            let mut nodes = vec![MemoryLayoutNode {
+                item_name,
+                typename: typename.clone(),
+                size: layout.size(),
+                offset: 0,
+                alignment: layout.align(),
+                parent_idx: -1,
+                children_start: -1,
+                children_len: 0,
+            }];
+            read_layout(&mut nodes, db, &ty, &layout, 0);
+
+            RecursiveMemoryLayout { nodes }
+        })
+        .ok()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::fixture;
+
+    fn make_memory_layout(ra_fixture: &str) -> Option<RecursiveMemoryLayout> {
+        let (analysis, position, _) = fixture::annotations(ra_fixture);
+
+        view_memory_layout(&analysis.db, position)
+    }
+
+    fn check_item_info<T>(node: &MemoryLayoutNode, item_name: &str, check_typename: bool) {
+        assert_eq!(node.item_name, item_name);
+        assert_eq!(node.size, core::mem::size_of::<T>() as u64);
+        assert_eq!(node.alignment, core::mem::align_of::<T>() as u64);
+        if check_typename {
+            assert_eq!(node.typename, std::any::type_name::<T>());
+        }
+    }
+
+    #[test]
+    fn view_memory_layout_none() {
+        assert!(make_memory_layout(r#"$0"#).is_none());
+        assert!(make_memory_layout(r#"stru$0ct Blah {}"#).is_none());
+    }
+
+    #[test]
+    fn view_memory_layout_primitive() {
+        let ml = make_memory_layout(
+            r#"
+fn main() {
+    let foo$0 = 109; // default i32
+}
+"#,
+        )
+        .unwrap();
+
+        assert_eq!(ml.nodes.len(), 1);
```

Comment:
Would it be better to write a to_string method and use `expect_test` here? It seems that these are prone to change due changes in rustc side so snapshot testing can help here (you tried to use `#[repr(C)]` for that, but it wouldn't work for e.g. tuples). The to_string method can just print each node with id of its children, but if you make it human readable we can probably also expose it to terminal based editors.

Reply:
I think all of the tests are independent from rustc layout variance stuffs to my knowledge, though the downside to this is that it doesn't test tuple members or rustc struct layouts. I think one could use an offset_of macro to test them maybe, but not sure about that and idk if rust can do that in a not weird and hacky way. 
As for to_string stuffs not really sure how that would change the problem? 
For the type 
```rust
struct AB {
  a: u32, 
  b: u32
}
``` 
rustc could lay it out as either (incomplete stringification as example):
```
AB: size 8, offset 0
  a: size 4, offset 0
  b: size 4, offset 4

or 

AB: size 8, offset 0
  b: size 4, offset 0
  a: size 4, offset 4
```
Even if it was reordered so that the fields were always consistently ordered the offsets would be different.

Having an in-built to_string method for terminals sounds like a good idea but since I don't use one not sure what design requirements it would need to consider.

Reply:
> rustc could lay it out as either (incomplete stringification as example):

The situation is not that bad. Rustc output is deterministic (even randomized layout which we don't support accepts a seed) but unstable. So it wouldn't change unless we bump the `rustc_abi` dependency crate version. Using to_string and [expect-test](https://crates.io/crates/expect-test) we can run `env UPDATE_EXPECT=1 cargo test` and update all these tests with minimal effort.

> I don't use one not sure what design requirements it would need to consider.

I don't know either, but I guess any string with a reasonable width should work? Maybe @nemethf can help us about this.

Reply:
d0df00d reworks all the tests to use expect-test, I wrote a pretty simple to_string impl, though its not super useful as a visualization tool so I didn't expose it through the api (not really sure it should be since its extra work for those who don't use it, and probably editors will want to take into account users' preferences for tab/space, colorization, etc? 🤷‍♀️).

Reply:
> Maybe @nemethf can help us about this.

Unfortunately, I cannot.  (I read this and other pull requests from the perspective of an LSP-client developer.)